### PR TITLE
Implement ConvtoBinaryConv transformation

### PR DESCRIPTION
**Pull request by mateusztabaka**
Ticket: 46687

### CODE REVIEW

File: inference-engine/src/transformations/src/transformations/common_optimizations/common_optimizations.cpp [hunk 1]
```diff
@@ -78,6 +80,7 @@ bool ngraph::pass::CommonOptimizations::run_on_function(std::shared_ptr<ngraph::
     eliminations->add_matcher<ngraph::pass::NopElimination>(); // may introduce fake dynamism
     eliminations->set_name("ngraph::pass::CommonEliminations");
 
+    manager.register_pass<ngraph::pass::BinarizeWeights>();
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/common_optimizations.cpp:83]: Why you decided to call it before ConstantFolding. Does it depends on it?

**Review by ilyachur — APPROVED** [review]

File: inference-engine/src/transformations/src/transformations/common_optimizations/common_optimizations.cpp (hunk 1 above)
Comment (mateusztabaka) [reply inference-engine/src/transformations/src/transformations/common_optimizations/common_optimizations.cpp:83]: first version of this pass heavily depended on CF, but that's not the case anymore so I can probably move that closer to ConvToBinaryConv transformation.

**Review by eshoguli — COMMENTED** [review]
No issues from LPT side. 
Note, please, some comment from LPT: if you separate binary FQ decomposition on activations and binary Convolution handling to two different transformations then you can easily handle  FQ on activations with several consumers

**Review by postrational — APPROVED** [review]

File: inference-engine/src/transformations/include/transformations/op_conversions/convert_reduce_to_pooling.hpp [hunk 2]
```diff
@@ -74,6 +74,11 @@ ngraph::matcher_pass_callback ConvertReduceBase::convert_reduce_to_pooling() {
 
         auto input = reduce->input_value(0);
 
+        // if input to Reduce op is a constant, it should be constant folded instead
+        if (std::dynamic_pointer_cast<ngraph::opset1::Constant>(input.get_node_shared_ptr())) {
```
[the remark this answers is not in the document]
Comment (GlebKazantaev) [reply inference-engine/src/transformations/include/transformations/op_conversions/convert_reduce_to_pooling.hpp:78]: If some transformation produce Reduce that must be constant folded then we need to call ConstantFolding pass but not add checks to other passes.

**Review by GlebKazantaev — CHANGES_REQUESTED** [review]

**Comment by GlebKazantaev** [comment]
In general LGTM. 
@jane-intel could you please take a look?

File: inference-engine/src/transformations/include/transformations/op_conversions/convert_reduce_to_pooling.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (mateusztabaka) [reply inference-engine/src/transformations/include/transformations/op_conversions/convert_reduce_to_pooling.hpp]: I removed that check and I'm using get_constant_from_source for ReduceSum

File: inference-engine/src/transformations/include/transformations/common_optimizations/binarize_weights.hpp [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+} // namespace pass
+} // namespace ngraph
+
+/**
+ * @ingroup ie_transformation_common_api
+ * @brief This transformation converts weights to -1/+1 form
+ * and applies normalization factors to output low/high and after Convolution.
+ * For example, following graph
+ *
+ *         .... ....  out_low  out_high           weights ..    ..  out_low out_high
```
[the remark this answers is not in the document]
Comment (andrew-zaytsev) [reply inference-engine/src/transformations/include/transformations/common_optimizations/binarize_weights.hpp:24]: No, we don't generate API docs for cpp/c files in the **src** folder, but only for headers in the **include** directory

**Review by jane-intel — APPROVED** [review]

**Review by GlebKazantaev — APPROVED** [review]

**Pull request merged**